packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -245,6 +290,51 @@ class StripeHelper {
       ],
     });
   }
+
+  /**
+   * Formats Stripe subscriptions for a customer into an appropriate response.
+   *
+   * @param {Subscriptions} subscriptions Subscriptions to finesse
+   * @returns {Promise<object[]>} Formatted list of subscriptions.
+   */
+  async subscriptionsToResponse(subscriptions) {
+    const subs = [];
+    for (const sub of subscriptions.data) {
+      let failure_code, failure_message;
+      // If this is a charge-automatically payment that is incomplete, attempt
+      // to get details of why it failed. The caller should expand the last_invoice
+      // calls by passing ['data.subscriptions.data.latest_invoice'] to `fetchCustomer`
+      // as the `expand` argument or this will not fetch the failure code/message.
+      if (
+        sub.status === 'incomplete' &&
+        typeof sub.latest_invoice !== 'string' &&
+        sub.collection_method === 'charge_automatically' &&
+        typeof sub.latest_invoice.charge === 'string'
+      ) {
+        const charge = await this.stripe.charges.retrieve(
+          sub.latest_invoice.charge
+        );
+        failure_code = charge.failure_code;
+        failure_message = charge.failure_message;
+      }
+      // FIXME: Note that the plan is only set if the subscription contains a single
+      //        plan. Multiple product support will require changes here to fetch all
+      //        plans for this subscription.
+      subs.push({
+        current_period_end: sub.current_period_end,
+        current_period_start: sub.current_period_start,
+        cancel_at_period_end: sub.cancel_at_period_end,
+        end_at: sub.ended_at,
+        plan_name: sub.plan.nickname,
+        plan_id: sub.plan.id,
+        status: sub.status,
+        subscription_id: sub.id,
+        failure_code,
+        failure_message,
+      });
+    }
+    return subs;
+  }
```

Comment:
Style preference comment: I prefer `return subscriptions.data.map(sub => { ... })`

Reply:
This code is doing an `await` in it though, which didn't seem to play nice with map afaik.

Reply:
Oh right!  This is the second time today I tripped on awaits in loops in reviews. 😛